### DISCUSSION TITLE
Enhance comparison by clarifying architecture models

### DISCUSSION THREAD

**Discussion by lcobucci** [TechEmpower/FrameworkBenchmarks#8553] [General]
While implementing improvements (#8550) and having discussions with some folks, I observed that we can increase the usefulness of the benchmark model by better categorising the architecture model used (instead of simply the framework).

Let's take PHP frameworks as example, the benchmark compares the mainstream model (share nothing architecture), with "native" async via event-loop, with "hidden" async with extensions, with "process/worker managers" based things.
That makes it difficult to separate "framework overhead" from "architectural constraints".

Expanding the perspective to other languages, causes this "apple vs pears" effects even more visible (we add "multi-process" vs "multi-thread" vs "multi virtual thread" to the mix).

Other discussions are pointing to the same thing but (IMHO) attacking the symptom and not the root cause.

My proposal would be to provide additional details regarding the architecture model on the implementations, which would also allow us to see the different consequences of using "framework z" against different architectural models.

**Comment by volyrique** [comment]
I understand where you are coming from, but are those architectural concepts really transferable, especially between different languages? It might be challenging to classify implementations consistently and uniformly.

**Reply by lcobucci** [reply]
I don't see why they wouldn't... Can you elaborate?

**Reply by volyrique** [reply]
Well, take "threads" as an example - in one environment the term may mean entities that are scheduled by the Linux kernel, while in another it may mean what is commonly known as [fibers](https://en.wikipedia.org/wiki/Fiber_(computer_science)) (or green threads). A third group of runtimes may actually implement [M:N threading](https://en.wikipedia.org/wiki/Thread_(computing)#M:N_(hybrid_threading)), which is like a hybrid of the other ones. So, if you say that an implementation is multi-threaded, for example, what do you mean exactly?!

> ... the benchmark compares the mainstream model (share nothing architecture), with "native" async via event-loop, with "hidden" async with extensions, with "process/worker managers" based things.

As the author of a C implementation, `h2o`, I don't really know what those terms mean. Sure, I might have some (vague) notion (about the last item, for instance), but that is not enough to determine if my code fits neatly into any of the categories. I suppose that this classification makes perfect sense if you limit yourself to a single language (i.e. PHP), but your proposal is clearly broader.

One way I could see it working is if the architecture is described from the viewpoint of the operating system, but I suspect that this is not the level of detail that you are looking for - for example, it is possible that some of your categories might appear to be exactly the same to Linux.

**Reply by lcobucci** [reply]
Thanks for providing more details @volyrique (and sorry for my huge delay :sweat_smile:)!
I understand your concern and believe we can find something that can make sense in a language-agnostic way.

I thought about the following matrix of parameters to define them:

* Webserver mode: external (using an external process - Nginx, Apache, etc); embedded (app incubated in the webserver runtime - Apache, Caddy, etc); user-land (implemented in the language)
* Process Management mode: single process; multi processes
* Threading: none; kernel-level threading; user-level threading; fibers (green/virtual threads); hybrid threading

Not sure how confusing it might be... I'm just concerned with the idea of comparing, for example, Symfony+Nginx+PHP-FPM with Symfony+[PPM](https://github.com/php-pm/php-pm) (a multi-processed web server implemented in PHP using fibers) with Symfony+[FrankenPHP](https://frankenphp.dev/docs/) (PHP embedded in a Go webserver).

I'm sure, though, that all supported frameworks would possibly have different results on the benchmark depending on the available architectural categories.

**Comment by joanhey** [comment]
Exist a long list of options, and sometimes they are blurred or even have various. For example, FrankenPHP use a different model if it's working in worker mode or not. And most options, less than 2% normal users will understand it.

> Not sure how confusing it might be... I'm just concerned with the idea of comparing, for example, Symfony+Nginx+PHP-FPM with Symfony+[PPM](https://github.com/php-pm/php-pm) (a multi-processed web server implemented in PHP using fibers) with Symfony+[FrankenPHP](https://frankenphp.dev/docs/) (PHP embedded in a Go webserver).

That's the idea of the benchmark, compare frameworks using any architecture.
Each platform use a different architecture, and normally the platform name is add to the name.

### Symfony
![image](https://github.com/TechEmpower/FrameworkBenchmarks/assets/249085/a299b67a-110e-4395-9079-bef3e93f164b)


Latency
![image](https://github.com/TechEmpower/FrameworkBenchmarks/assets/249085/499b965a-34a6-456a-9422-1780acdaa73d)

Also we have the classification columns, than we can filter to show only for specific options. Type of framework, language, platform, server, ....
So we can compare only with selected platforms (architecture or stack). 

![image](https://github.com/TechEmpower/FrameworkBenchmarks/assets/249085/5e92bb59-dc23-4dc6-b68c-996d7c9a48d3)

We only need to filter to view only the preferred options.
![image](https://github.com/TechEmpower/FrameworkBenchmarks/assets/249085/d0117726-392c-46c1-9185-c22dbc09cfc2)

Example to only show the frameworks using fpm/fastcgi
![image](https://github.com/TechEmpower/FrameworkBenchmarks/assets/249085/5e11693e-7773-4b12-84c8-d23a6fe75d4f)

I tried for some years that the plaintext test show in the classification, the frameworks using http1.1 pipeline. And still is not added.
Http1.1 pipeline, it is not used by any browser and most proxies.

**Reply by lcobucci** [reply]
What I'm trying to convey is that I believe it'd be useful to have a more generic classification for things. 

For example: being able to compare nodejs and python solutions with php ones that follow the same architectural design.

**Comment by joanhey** [comment]
You can do it, with the filters !!!

**Comment by joanhey** [comment]
I prefer say stack (architecture)
Do you want nodjs, python and PHP with the same architecture.
Say the exact architecture, and I'll send you a link to that comparison!!

**Reply by lcobucci** [reply]
Those were just examples... the issue with the filters is that people need to know about the architectural details of the platforms in order to achieve what I'm describing.

Say someone wants to compare everything of all languages that is multi-process based with native http server, or everything that uses hybrid threading... idk

It's fine if that's not a problem in opinion, btw 😁

**Comment by joanhey** [comment]
Nodejs only have one architecture, Python have a lot, ... and PHP is new having new architectures from the classic fcgi shared nothing.

**Comment by joanhey** [comment]
It will be not easy for me, to try the correct architecture with python. But I'll try my best.

Edit: with PHP will be not easy too!!!

**Comment by joanhey** [comment]
First, when you find the people that understand it, please say me !!!

PD: thank you !!

**Comment by joanhey** [comment]
This benchmark have more than 400 frameworks, with more than 300 platforms, ....
What do you want exactly, and we'll try to help.

**Comment by joanhey** [comment]
Nobody is interest about the technical details, they only want use less servers for serve their users..

**Reply by lcobucci** [reply]
Looking at the architectural models allow us to comparing apples to apples and pears to pears, regardless of language or stack, which makes the benchmark more accurate in general.

Let's just say it's not a concern for now, then. Thanks for your input 👍

**Comment by joanhey** [comment]
https://www.techempower.com/benchmarks/#section=test&runid=502e3495-dda5-4232-a89a-7c08755fcf0e&hw=ph&test=db&l=v2qiv3-35r&f=6bk-27y5mo-mh35-2hwp34-0-0-5k36rk-282dfk-3s-mh34-8vn08w-2-e8-9zo6ww-0

Only PHP with any platform !!!
PD:  sorry, not all php frameworks and platforms, and with the new servers it'll change a lot !!!
